Makefile

```diff
@@ -1,119 +1,85 @@
-USE_FRAMEBUFFER := 1
-SMALL_PAGES := 0
-CFLAGS := -std=gnu99 \
-        -ffreestanding \
-        -O2 \
-        -Wall \
-        -Wextra \
-        -I src/include \
-        -I src/include/base \
-        -I src/include/kernel \
-        -I src/include/kernel/mem \
-        -I src/include/kernel/x86_64 \
-        -I src/include/kernel/hardware \
-        -I src/include/kernel/scheduling \
-        -I src/include/libc \
-        -I src/include/fs \
-        -I src/include/drivers/fs \
-        -I src/include/sys \
-        -mno-red-zone \
-        -mno-sse \
-        -mcmodel=large \
-        -DUSE_FRAMEBUFFER=$(USE_FRAMEBUFFER) \
-        -DSMALL_PAGES=$(SMALL_PAGES)
-
-TESTFLAGS := -std=gnu99 \
-        -I tests/include \
-        -I src/include \
-        -I src/include/base \
-        -I src/include/kernel/mem \
-        -I src/include/fs \
-        -I src/include/drivers/fs \
-        -I src/include/kernel \
-        -I src/include/kernel/x86_64 \
-        -I src/include/sys \
-        -DSMALL_PAGES=$(SMALL_PAGES) \
-        -D_TEST_=1
-
-TEST_FOLDER := tests
-C_DEBUG_FLAGS := -g \
-				-DDEBUG=1
-NASM_DEBUG_FLAGS := -g \
+include build/Config.mk
+include build/Common.mk
+
+# arch dependant
+ARCH_PREFIX := x86_64-elf
+ASM_COMPILER := nasm
+QEMU_SYSTEM := qemu-system-x86_64
+
+ASM_DEBUG_FLAGS := -g \
 					-F dwarf
-NASMFLAGS := -f elf64 \
+ASM_FLAGS := -f elf64 \
 		-D USE_FRAMEBUFFER=$(USE_FRAMEBUFFER) \
 		-D SMALL_PAGES=$(SMALL_PAGES)
-BUILD := build
-PRJ_FOLDERS := src
-FONT_FOLDERS := fonts
+
 DEBUG := 0
-VERBOSE_OUTPUT :=0
 
 SRC_C_FILES := $(shell find $(PRJ_FOLDERS) -type f -name "*.c")
 SRC_H_FILES := $(shell find $(PRJ_FOLDERS) -type f -name "*.h")
 SRC_ASM_FILES := $(shell find $(PRJ_FOLDERS) -type f -name "*.s")
 SRC_FONT_FILES := $(shell find $(FONT_FOLDERS) -type f -name "*.psf")
-OBJ_ASM_FILE := $(patsubst src/%.s, build/%.o, $(SRC_ASM_FILES))
-OBJ_C_FILE := $(patsubst src/%.c, build/%.o, $(SRC_C_FILES))
-OBJ_FONT_FILE := $(patsubst fonts/%.psf, build/%.o, $(SRC_FONT_FILES))
+OBJ_ASM_FILE := $(patsubst src/%.s, $(BUILD_FOLDER)/%.o, $(SRC_ASM_FILES))
+OBJ_C_FILE := $(patsubst src/%.c, $(BUILD_FOLDER)/%.o, $(SRC_C_FILES))
+OBJ_FONT_FILE := $(patsubst fonts/%.psf, $(BUILD_FOLDER)/%.o, $(SRC_FONT_FILES))
+
 default: build
 
 .PHONY: default build run clean debug tests gdb
 
-build: build/os.iso
+build: $(BUILD_FOLDER)/os.iso
 
 clean:
-	rm -rf build
+	rm -rf $(BUILD_FOLDER)
```

Comment:
For clean targets I like prepending a `-` before the command, this tells make to ignore the result. So if you accidentally run `make clean` twice in a row it will fail silently instead. Personal choice here though, not a recommended practice.